text/0000-decorators.md

```diff
@@ -0,0 +1,1230 @@
+
+The `@action` decorator _could_ warn users when it collides with a lifecycle
+hook. However, hooks may vary from class type to class type, which presents a
+design challenge. We could either:
+
+1. Allow classes to specify lifecycle hooks, and throw whenever `@action`
+   collides with a specified hook.
+2. Only throw on hooks that are shared across all classes, such as `init` and
+   `destroy`.
+3. Do nothing, and leave it to user's to know which lifecycle hooks exist.
+
+Specifying hooks for each class would be time consuming and could fall out of
+sync with the implementations. Throwing on "universal" hooks only is
+inconsistent, and could lead users to think that they are safe when they are
+not. This RFC suggests that we choose option 3 for this reason. When the actions
+namespace was introduced, lifecycle hooks like `destroy` and `click` were less
+commonly known and used (event listeners were not uncommon). Most Ember users
+know they exist now, and will be aware that implementing an action with the same
+name is not recommended. In addition, eslint rtules can be added to hint against
+these collisions.
+
+This also aligns with the future laid out by Glimmer components which will be
+discussed more thoroughly in upcoming RFCs, where actions will not need to be
+decorated at all, and will be plain methods:
+
+```js
+export default class ButtonComponent extends GlimmerComponent {
+  onClick() {
+    // handle click
+  }
+}
+```
+```hbs
+<button onclick={{action this.onClick}}>Click me!</button>
+```
+
+This pattern can be used today with legacy components and controllers, but
+routes still depend heavily on `sendAction` and actions being made available
+internally via the actions object, so for the time being, we need a way to mark
+methods as actions.
+
+#### Usage and API
+
+The API for this new decorator would be much simpler than the computed API,
+since it is only used as a decorator without parameters.
+
+```ts
+// Technically `action` is a function, but we can't type it transparently that way
+const action: MethodDecorator;
+```
+
+Attempting to pass any parameters to the decorator, or to apply the decorator to
+anything other than a class method, will throw an error.
+
+## How we teach this
+
+Teaching decorators is intrinsically tied to a wider shift in the Ember
+programming model - the Ember Octane edition. From a teaching perspective, this
+edition will be completely overhauling the guides and updating all of the best
+practices as they stand. New users should see native class syntax with
+decorators as the _default_, and should not ever have to write a legacy class
+or see an example for one.
+
+With this RFC, the majority of existing examples in the Ember guides will be
+updatable to native class syntax. The exception would be examples of legacy
+components, which would be addressed separately by updating the guides to
+Glimmer Components (proposed in a separate RFC). Otherwise, all examples in the
+guides should be updated.
```

Comment:
I'd go one step further, and suggest we support `Ember.Object.extend` we should have current-version documentation describing proper usage, at _least_ until `Ember.Object.extend` is deprecated. 

As you point out, this could be relegated to (and confined within) a "legacy programming model" section of the guides or something.